pkg/auth/authenticator/request/x509request/x509.go

```diff
@@ -0,0 +1,95 @@
+package x509request
+
+import (
+	"crypto/x509"
+	"net/http"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/errors"
+	"github.com/openshift/origin/pkg/auth/api"
+)
+
+// UserConversion defines an interface for extracting user info from a client certificate chain
+type UserConversion interface {
+	User(chain []*x509.Certificate) (api.UserInfo, bool, error)
+}
+
+// UserConversionFunc is a function that implements the UserConversion interface.
+type UserConversionFunc func(chain []*x509.Certificate) (api.UserInfo, bool, error)
+
+// User implements x509.UserConversion
+func (f UserConversionFunc) User(chain []*x509.Certificate) (api.UserInfo, bool, error) {
+	return f(chain)
+}
+
+// Authenticator implements request.Authenticator by extracting user info from verified client certificates
+type Authenticator struct {
+	opts x509.VerifyOptions
+	user UserConversion
+}
+
+// New returns a request.Authenticator that verifies client certificates using the provided
+// VerifyOptions, and converts valid certificate chains into api.UserInfo using the provided UserConversion
+func New(opts x509.VerifyOptions, user UserConversion) *Authenticator {
+	return &Authenticator{opts, user}
+}
+
+// AuthenticateRequest authenticates the request using presented client certificates
+func (a *Authenticator) AuthenticateRequest(req *http.Request) (api.UserInfo, bool, error) {
+	if req.TLS == nil {
+		return nil, false, nil
+	}
+
+	var errlist []error
+	for _, cert := range req.TLS.PeerCertificates {
+		chains, err := cert.Verify(a.opts)
+		if err != nil {
+			errlist = append(errlist, err)
+			continue
+		}
+
+		for _, chain := range chains {
+			user, ok, err := a.user.User(chain)
+			if err != nil {
+				errlist = append(errlist, err)
+				continue
+			}
+
+			if ok {
+				return user, ok, err
```

Comment:
I may have missed it, but are these names being properly provider scoped?  In the past, we used an identity mapper to take a name from something like a PasswordAuthenticator and scope it down.